src/prelude/timers.mo

```diff
@@ -0,0 +1,149 @@
+var @timers : ?@Node = null;
+func @prune(n : ?@Node) : ?@Node = switch n {
+    case null null;
+    case (?n) {
+        if (n.expire == 0) {
+            @prune(n.dopo) // by corollary
+        } else {
+            ?{ n with ante = @prune(n.ante); dopo = @prune(n.dopo) }
+        }
+    }
+};
+
+func @nextExpiration(n : ?@Node) : Nat64 = switch n {
+    case null 0;
+    case (?n) {
+        var exp = @nextExpiration(n.ante); // TODO: use the corollary for expire == 0
+        if (exp == 0) {
+            exp := n.expire;
+            if (exp == 0) {
+                exp := @nextExpiration(n.dopo)
+            }
+        };
+        exp
+    }
+};
+
+// Function called by backend to run eligible timed actions.
+// DO NOT RENAME without modifying compilation.
+func @timer_helper() : async () {
+    func Array_init<T>(len : Nat,  x : T) : [var T] {
+        (prim "Array.init" : <T>(Nat, T) -> [var T])<T>(len, x)
+    };
+    let now = (prim "time" : () -> Nat64)();
+    let exp = @nextExpiration @timers;
+    let prev = (prim "global_timer_set" : Nat64 -> Nat64) exp;
+
+    if (exp == 0) {
+        return
+    };
+
+    var gathered = 0;
+    let thunks : [var ?(() -> async ())] = Array_init(10, null); // we want max 10
+
+    func gatherExpired(n : ?@Node) = switch n {
+        case null ();
+        case (?n) {
+            gatherExpired(n.ante);
+            if (n.expire > 0 and n.expire <= now and gathered < thunks.size()) {
+                thunks[gathered] := ?(n.job);
+                switch (n.delay) {
+                    case (?delay) {
+                        // re-add the node
+                        let expire = n.expire + delay;
+                        // N.B. insert only works on pruned nodes
+                        func reinsert(m : ?@Node) : @Node = switch m {
+                            case null ({ n with var expire; ante = null; dopo = null });
+                            case (?m) {
+                                assert m.expire != 0;
+                                if (expire < m.expire) ({ m with ante = ?reinsert(m.ante) })
+                                else ({ m with dopo = ?reinsert(m.dopo) })
+                            }
+                        };
+                        @timers := ?reinsert(@prune(@timers));
+                    };
+                    case _ ()
+                };
+                n.expire := 0;
+                gathered += 1;
+            };
+            gatherExpired(n.dopo);
+        }
+    };
+
+    gatherExpired(@timers);
+
+    let futures : [var ?(async ())] = Array_init(thunks.size(), null);
+    for (k in thunks.keys()) {
+        futures[k] := switch (thunks[k]) { case (?thunk) ?thunk(); case _ null };
+    };
+
+    for (f in futures.vals()) {
+        switch f { case (?f) try await f catch _ { }; case _ () }
+    };
+};
+
+var @lastId = 0;
+
+func @setTimer(delaySecs : Nat64, recurring : Bool, job : () -> async ()) : (id : Nat) {
```

Comment:
We have `set_timer`() and `set_timer_interval()` implemented in [Rust CDK](https://github.com/dfinity/cdk-rs/pull/342/files#diff-b199dfe6f7df9ad54346b557c68334c9b3a75f22784b195d6696f80586469bffR172).

Do you think it makes sense to provide the same API for Motoko, i.e. `setTimer()` and `setTimerInterval()` for recurring tasks?

Reply:
I don't see why one would want two setting routines. I think the `setTimer`/`cancelTimer` pair is more economical.

Reply:
The thread is marked as outdated, but I guess the comment still stands: we better to align with Rust CDK and either implement in both languages `setTImer` + `setTimerInterval` or `setTimer(..., recurring)`...